src/tir/transforms/unify_thread_binding.cc

```diff
@@ -0,0 +1,143 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file unify_thread_binding.cc
+ */
+
+#include <tvm/arith/analyzer.h>
+#include <tvm/tir/analysis.h>
+#include <tvm/tir/stmt_functor.h>
+#include <tvm/tir/transform.h>
+
+#include "ir_utils.h"
+
+namespace tvm {
+namespace tir {
+
+/*!
+ * \brief A mutator which searches AttrStmts of thread bindings and changes the `node` field IterVar
+ * of the AttrStmts, so that for one kind of thread binding, all such thread bindings use the same
+ * IterVar
+ */
+class ThreadBindingUnifier : public StmtExprMutator {
+ public:
+  static Stmt Unify(const Stmt& stmt) { return ThreadBindingUnifier().VisitStmt(stmt); }
+
+ private:
+  Stmt VisitStmt_(const AttrStmtNode* attr) final {
+    // If this AttrStmt is not thread binding attribute, return as usual.
+    if (attr->attr_key != attr::thread_extent && attr->attr_key != attr::virtual_thread) {
+      return StmtMutator::VisitStmt_(attr);
+    }
+
+    // Step 1. Fetch the old IterVar and the thread tag.
+    IterVar old_iter_var = Downcast<IterVar>(attr->node);
+    IterVar new_iter_var{nullptr};
+    const String& thread_tag = old_iter_var->thread_tag;
+
+    // Step 2: Increase `thread_block_depth_` if the thread tag starts with "blockIdx". If the
+    // thread block depth is 0 before the increasement, it means we are entering a new kernel, and
+    // therefore we need to make `thread_tag2iter_var_map_` empty, as different kernels can have
+    // thread axes with different extents.
+    if (std::string(thread_tag).substr(0, 9) == "blockIdx.") {
+      if (!thread_block_depth_) {
+        thread_tag2iter_var_map_.clear();
+      }
+      ++thread_block_depth_;
+    }
+
+    // Step 3. See if an IterVar for this kind of thread binding was created before. If so, we use
+    // the created IterVar. Otherwise, we create a new IterVar for this thread binding and store the
+    // IterVar in mapping `thread_tag2iter_var_map_`.
+    Map<String, IterVar>::iterator it = thread_tag2iter_var_map_.find(thread_tag);
+    if (it != thread_tag2iter_var_map_.end()) {
+      new_iter_var = (*it).second;
+      arith::Analyzer ana;
+      CHECK(ana.CanProveEqual(old_iter_var->dom->extent, (*it).second->dom->extent))
```

Comment:
I create a temporary variable here because if I wrote
```c++
CHECK(arith::Analyzer().CanProveEqual(old_iter_var->dom->extent, (*it).second->dom->extent))
```
, it throws `TVMError` instead of `ValueError`. The `TVMError` is as the following:
```
TVMError: Check failed: (arith: :Analyzer().CanProveEqual(old_iter_var->dom->extent, (*it).second->dom->extent)) is false: ValueError: All loops that are bound to `threadIdx.x` should have the same extent. However, there are two loops with extent 128 and 64, which are not equal
```
I guess the `arith: :Analyzer()` (note the non-consecutive colons) triggers the issue.

@junrushao1994 See if you have any comments? Keep using a temporary variable, or throw `TVMError`, or fix the non-consecutive-colons issue?

Reply:
First, please have an analyzer in the class so that you don’t need to create a new one every time.

Second, if you wanted to convert this to ValueError, just write it down in the error message:

CHECK(…) << “ValueError: blah”;

the system will take care of it for you :-)

Reply:
> First, please have an analyzer in the class so that you don’t need to create a new one every time.

Thanks! Will do it now.